signac/common/config.py

```diff
@@ -20,65 +20,64 @@ def _get_project_config_fn(root):
     return os.path.abspath(os.path.join(root, PROJECT_CONFIG_FN))
 
 
-def _get_config_dirname(fn):
-    # We could use pathlib's `.paths` attribute to remove the trailing project
-    # config filename, but that has significant performance
-    # implications: this takes ~200 ns, whereas
-    # os.path.join(*(pathlib.PosixPath(fn).parts[:-2])) takes ~6 us, almost a
-    # 20x slowdown.
-    return fn.replace(os.sep + PROJECT_CONFIG_FN, "")
+def _contains_config_file(root):
+    """Determine if the root directory contains a signac configuration file.
 
+    Parameters
+    ----------
+    root : str
+        Path to search. Uses ``os.getcwd()`` if None (Default value = None).
 
-def _search_local(root):
+    Returns
+    --------
+    str or None
+        The path to the configuration file if one is found, otherwise None.
+    """
     fn_ = _get_project_config_fn(root)
     if os.path.isfile(fn_):
-        yield fn_
+        return root
+    return None
 
 
-def _search_tree(root=None):
-    """Locates signac configuration files in a directory hierarchy.
+def _locate_config_dir(root):
+    """Locates root directory containing a signac configuration file in a directory hierarchy.
 
     Parameters
     ----------
     root : str
-        Path to search. Uses ``os.getcwd()`` if None (Default value = None).
+        Starting path to search.
 
+    Returns
+    --------
+    str or None
+        The root directory containing the configuration file if one is found, otherwise None.
     """
-    if root is None:
-        root = os.getcwd()
     while True:
-        yield from _search_local(root)
+        if _contains_config_file(root):
+            return root
         up = os.path.abspath(os.path.join(root, ".."))
```

Comment:
We should create an absolute path once, before entering the `while True:`, then use `os.path.dirname` each time we want to go up. Performance is better if we (1) use `dirname` to go up instead of a join with `".."` and (2) call `abspath` fewer times.

```python
root = os.path.abspath(root)
while True:
    # ...
    up = os.path.dirname(root)
    # ...
```

Reply:
https://github.com/bdice/python_microbenchmarks/blob/main/microbenchmarks/os_path_dirname.py

Relative cost of each function call:
```
dirname: 0.0445243
join: 0.0719918
abs-dirname: 0.176682
abs-join: 0.234717
```

Reply:
Good call.